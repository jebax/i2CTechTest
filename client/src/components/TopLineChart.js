import React from 'react'
import PropTypes from 'prop-types'
import {
  XYPlot,
  XAxis,
  YAxis,
  VerticalGridLines,
  HorizontalGridLines,
  VerticalBarSeries,
  ChartLabel
} from 'react-vis'

const TopLineChart = ({ spendData, unitsData, visitsData, totalCustsData }) => {
  return (
    <XYPlot
      height={300}
      width={600}
      xType={'ordinal'}
    >
      <XAxis />
      <YAxis />
      <VerticalGridLines />
      <HorizontalGridLines />
      <ChartLabel
        text="Uplift (%)"
        className="top-axis-label"
        includeMargin={false}
        xPercent={0.06}
        yPercent={0.06}
        style={{
          transform: 'rotate(-90)',
          textAnchor: 'end'
        }}
      />
      <VerticalBarSeries
        data={spendData}
      />
      <VerticalBarSeries
        data={unitsData}
      />
      <VerticalBarSeries
        data={visitsData}
      />
      <VerticalBarSeries
        data={totalCustsData}
      />
    </XYPlot>
  )
}

TopLineChart.propTypes = {
  spendData: PropTypes.array.isRequired,
  unitsData: PropTypes.array.isRequired,
  visitsData: PropTypes.array.isRequired,
  totalCustsData: PropTypes.array.isRequired
}

export default TopLineChart
