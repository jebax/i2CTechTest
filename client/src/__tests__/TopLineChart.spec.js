import React from 'react'
import { shallow } from 'enzyme'
import TopLineChart from '../components/TopLineChart'

describe('Top line chart', () => {
  var topLineChart
  var spendData
  var unitsData
  var visitsData
  var totalCustsData

  beforeAll(() => {
    spendData = [{ x: 'Aisle', y: '0.01' }]
    unitsData = [{ x: 'Aisle', y: '0.02' }]
    visitsData = [{ x: 'Aisle', y: '0.03' }]
    totalCustsData = [{ x: 'Aisle', y: '0.04' }]

    topLineChart = shallow(
      <TopLineChart
        spendData={spendData}
        unitsData={unitsData}
        visitsData={visitsData}
        totalCustsData={totalCustsData}
      />
    )
  })

  it('renders correctly', () => {
    expect(topLineChart).toMatchSnapshot()
  })

  it('renders a react-vis chart', () => {
    expect(topLineChart.find('XYPlot').length).toEqual(1)
  })

  it('displays ordinal data on its x-axis', () => {
    const plot = topLineChart.find('XYPlot')
    expect(plot.props().xType).toEqual('ordinal')
  })

  it('renders four VerticalBarSeries', () => {
    expect(topLineChart.find('VerticalBarSeries').length).toEqual(4)
  })

  it('should display the spend data in the first bar series', () => {
    const firstSeries = topLineChart.find('VerticalBarSeries').first()

    expect(firstSeries.props().data).toEqual(spendData)
  })

  it('should display the units data in the second bar series', () => {
    const secondSeries = topLineChart.find('VerticalBarSeries').at(1)

    expect(secondSeries.props().data).toEqual(unitsData)
  })

  it('should display the visits data in the third bar series', () => {
    const thirdSeries = topLineChart.find('VerticalBarSeries').at(2)

    expect(thirdSeries.props().data).toEqual(visitsData)
  })

  it('should display the total customers data in the fourth bar series', () => {
    const fourthSeries = topLineChart.find('VerticalBarSeries').at(3)

    expect(fourthSeries.props().data).toEqual(totalCustsData)
  })

  it('renders a chart label for the percentage uplift', () => {
    const label = topLineChart.find('ChartLabel')

    expect(label.props().text).toEqual('Uplift (%)')
  })

  it('renders a DiscreteColorLegend item', () => {
    expect(topLineChart.find('DiscreteColorLegendItem').length).toEqual(1)
  })
})
