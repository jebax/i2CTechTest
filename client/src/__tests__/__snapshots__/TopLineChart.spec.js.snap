// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top line chart renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TopLineChart
    spendData={
      Array [
        Object {
          "x": "Aisle",
          "y": "0.01",
        },
      ]
    }
    totalCustsData={
      Array [
        Object {
          "x": "Aisle",
          "y": "0.04",
        },
      ]
    }
    unitsData={
      Array [
        Object {
          "x": "Aisle",
          "y": "0.02",
        },
      ]
    }
    visitsData={
      Array [
        Object {
          "x": "Aisle",
          "y": "0.03",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <XAxis
          attr="x"
          attrAxis="y"
          orientation="bottom"
        />,
        <YAxis
          attr="y"
          attrAxis="x"
          orientation="left"
        />,
        <VerticalGridLines
          attr="x"
          direction="vertical"
        />,
        <HorizontalGridLines
          attr="y"
          direction="horizontal"
        />,
        <ChartLabel
          className="top-axis-label"
          includeMargin={false}
          style={
            Object {
              "textAnchor": "end",
              "transform": "rotate(-90)",
            }
          }
          text="Uplift (%)"
          xPercent={0.06}
          yPercent={0.06}
        />,
        <VerticalBarSeries
          className=""
          data={
            Array [
              Object {
                "x": "Aisle",
                "y": "0.01",
              },
            ]
          }
          stack={false}
          style={Object {}}
        />,
        <VerticalBarSeries
          className=""
          data={
            Array [
              Object {
                "x": "Aisle",
                "y": "0.02",
              },
            ]
          }
          stack={false}
          style={Object {}}
        />,
        <VerticalBarSeries
          className=""
          data={
            Array [
              Object {
                "x": "Aisle",
                "y": "0.03",
              },
            ]
          }
          stack={false}
          style={Object {}}
        />,
        <VerticalBarSeries
          className=""
          data={
            Array [
              Object {
                "x": "Aisle",
                "y": "0.04",
              },
            ]
          }
          stack={false}
          style={Object {}}
        />,
      ],
      "className": "",
      "height": 300,
      "width": 600,
      "xType": "ordinal",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "attr": "x",
          "attrAxis": "y",
          "orientation": "bottom",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "attr": "y",
          "attrAxis": "x",
          "orientation": "left",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "attr": "x",
          "direction": "vertical",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "attr": "y",
          "direction": "horizontal",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "top-axis-label",
          "includeMargin": false,
          "style": Object {
            "textAnchor": "end",
            "transform": "rotate(-90)",
          },
          "text": "Uplift (%)",
          "xPercent": 0.06,
          "yPercent": 0.06,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "data": Array [
            Object {
              "x": "Aisle",
              "y": "0.01",
            },
          ],
          "stack": false,
          "style": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "data": Array [
            Object {
              "x": "Aisle",
              "y": "0.02",
            },
          ],
          "stack": false,
          "style": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "data": Array [
            Object {
              "x": "Aisle",
              "y": "0.03",
            },
          ],
          "stack": false,
          "style": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "data": Array [
            Object {
              "x": "Aisle",
              "y": "0.04",
            },
          ],
          "stack": false,
          "style": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <XAxis
            attr="x"
            attrAxis="y"
            orientation="bottom"
          />,
          <YAxis
            attr="y"
            attrAxis="x"
            orientation="left"
          />,
          <VerticalGridLines
            attr="x"
            direction="vertical"
          />,
          <HorizontalGridLines
            attr="y"
            direction="horizontal"
          />,
          <ChartLabel
            className="top-axis-label"
            includeMargin={false}
            style={
              Object {
                "textAnchor": "end",
                "transform": "rotate(-90)",
              }
            }
            text="Uplift (%)"
            xPercent={0.06}
            yPercent={0.06}
          />,
          <VerticalBarSeries
            className=""
            data={
              Array [
                Object {
                  "x": "Aisle",
                  "y": "0.01",
                },
              ]
            }
            stack={false}
            style={Object {}}
          />,
          <VerticalBarSeries
            className=""
            data={
              Array [
                Object {
                  "x": "Aisle",
                  "y": "0.02",
                },
              ]
            }
            stack={false}
            style={Object {}}
          />,
          <VerticalBarSeries
            className=""
            data={
              Array [
                Object {
                  "x": "Aisle",
                  "y": "0.03",
                },
              ]
            }
            stack={false}
            style={Object {}}
          />,
          <VerticalBarSeries
            className=""
            data={
              Array [
                Object {
                  "x": "Aisle",
                  "y": "0.04",
                },
              ]
            }
            stack={false}
            style={Object {}}
          />,
        ],
        "className": "",
        "height": 300,
        "width": 600,
        "xType": "ordinal",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "attr": "x",
            "attrAxis": "y",
            "orientation": "bottom",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "attr": "y",
            "attrAxis": "x",
            "orientation": "left",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "attr": "x",
            "direction": "vertical",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "attr": "y",
            "direction": "horizontal",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "top-axis-label",
            "includeMargin": false,
            "style": Object {
              "textAnchor": "end",
              "transform": "rotate(-90)",
            },
            "text": "Uplift (%)",
            "xPercent": 0.06,
            "yPercent": 0.06,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "data": Array [
              Object {
                "x": "Aisle",
                "y": "0.01",
              },
            ],
            "stack": false,
            "style": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "data": Array [
              Object {
                "x": "Aisle",
                "y": "0.02",
              },
            ],
            "stack": false,
            "style": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "data": Array [
              Object {
                "x": "Aisle",
                "y": "0.03",
              },
            ],
            "stack": false,
            "style": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "data": Array [
              Object {
                "x": "Aisle",
                "y": "0.04",
              },
            ],
            "stack": false,
            "style": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
